compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -906,11 +912,169 @@ private boolean checkSimilarMatchPatterns(BLangMatchPattern firstPattern, BLangM
             case MAPPING_MATCH_PATTERN:
                 return checkSimilarMappingMatchPattern((BLangMappingMatchPattern) firstPattern,
                         (BLangMappingMatchPattern) secondPattern);
+            case ERROR_MATCH_PATTERN:
+                return checkSimilarErrorMatchPattern((BLangErrorMatchPattern) firstPattern,
+                        (BLangErrorMatchPattern) secondPattern);
             default:
                 return false;
         }
     }
 
+    private boolean checkSimilarErrorMatchPattern(BLangErrorMatchPattern firstErrorMatchPattern,
+                                                  BLangErrorMatchPattern secondErrorMatchPattern) {
+        if (!checkSimilarErrorTypeReference(firstErrorMatchPattern.errorTypeReference,
+                secondErrorMatchPattern.errorTypeReference)) {
+            return false;
+        }
+
+        if (!checkSimilarErrorMessagePattern(firstErrorMatchPattern.errorMessageMatchPattern,
+                secondErrorMatchPattern.errorMessageMatchPattern)) {
+            return false;
+        }
+
+        if (!checkSimilarErrorCauseMatchPattern(firstErrorMatchPattern.errorCauseMatchPattern,
+                secondErrorMatchPattern.errorCauseMatchPattern)) {
+            return false;
+        }
+
+        if (!checkSimilarErrorFieldMatchPatterns(firstErrorMatchPattern.errorFieldMatchPatterns,
+                secondErrorMatchPattern.errorFieldMatchPatterns)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean checkSimilarErrorTypeReference(BLangUserDefinedType firstErrorTypeRef,
+                                                   BLangUserDefinedType secondErrorTypeRef) {
+        if ((firstErrorTypeRef != null && secondErrorTypeRef == null)
+                || (firstErrorTypeRef == null && secondErrorTypeRef != null)) {
+            return false;
+        }
+        if (firstErrorTypeRef == null) {
+            return true;
+        }
+        return firstErrorTypeRef.typeName.value.equals(secondErrorTypeRef.typeName.value);
```

Comment:
Shouldn't we check types instead?

If we have 
```ballerina
type ErrorOne error<record {int code;}>;
type ErrorTwo error<record {int code;}>;

public function func(error err) {
    match err {
        error ErrorOne() => {
        }
        error ErrorTwo() => {
        }
    }
}
```
the second pattern is unreachable, right?